manuals/wrcinfo.1
.Dd June 25, 2026
.Dt WRCINFO 1
.Os
.Sh NAME
.Nm wrcinfo
.Nd determines information about a Windows Resource (RC) stream of the .rsrc \
section of a MZ, PE/COFF executable
.Sh SYNOPSIS
.Nm wrcinfo
.Op Fl c Ar codepage
.Op Fl hHvV
.Ar source
.Sh DESCRIPTION
.Nm wrcinfo
is a utility to determine information about a Windows Resource (RC) stream of \
the .rsrc section of a MZ, PE/COFF executable
.Pp
.Nm wrcinfo
is part of the
.Nm libwrc
package.
.Nm libwrc
is a library to access the Windows Resource Compiler (WRC) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar codepage
codepage of ASCII strings, options: ascii, windows-874, windows-932, \
windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 \
(default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 \
or windows-1258
.It Fl h
shows this help
.It Fl H
shows the resource hierarchy
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# wrcinfo adtschema.dll
wrcinfo 20260625
.sp
Windows Resource (RC) information:
.sp
Resources:
	11 (MESSAGETABLE)
	14 (VERSIONINFO)
	WEVT_TEMPLATE
	MUI
.sp
Version resource:
	number of languages	: 1
.sp
	language identifier	: 0x00000409 (en-US)
	file version		: 6.0.6000.16386
	product version		: 6.0.6000.16386
.sp
MUI resource:
	number of languages	: 1
.sp
	language identifier	: 0x00000409 (en-US)
	file type		: 0x00000011
	main name		: WEVT_TEMPLATE
	MUI name		: MUI
	language		: en-US
	fallback language	: en-US
.sp
Message-table resource:
	number of languages	: 1
.sp
	language identifier	: 0x00000409 (en-US)
	number of messages	: 399
.sp
	message			: 0
	identifier		: 0x70003000
	data			: Security State Change
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libwrc/issues
.Sh COPYRIGHT
Copyright (C) 2011-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* escape special characters in string table and message table output

* pywrc:
  - MUI check output of pywrc_mui_get_main_name amd pywrc_mui_get_mui_name
  - add version type support (complete pywrc_version.c)
  - add iterator objects for a message table of a specific language
  - add iterator objects for strings of a specific language
* MUI
  - match muirct output
* VERSION
  - check if libwrc_version_values_read_string works
  - check if multi language version is now supported
* virtual address can be outside .rsrc section (in UPX packed files?)
  - location of unpacked runtime data ?
* make sure data types are large enough
* API:
  - add get resource name functions
* create gerenic solution to store resource data ?
* detect invalid padding
* use libfcache/libfdata in libwrc ? otherwise remove it
* language table
  - cache last language ?

Tools:
* wrcinfo:
  - what about PE/COFF files with multiple .rsrc sections?
* wrcmount
  - create, needs resource item API

20140928
* see `git log' for more recent change log
* removed README.macosx
* changes for project site move

20140803
* bug fix in Python-bindings

20140328
* updated dependencies
* Worked on Python bindings

20140209
* worked on format support

20140128
* worked on stripping trailing 0-byte values from message table strings
* improved handling of version resource

20140127
* Worked on Python bindings
  - clean up and bug fixes
  - added MUI type support

20140112
* small fixes for file IO handle opened in library changes

20140105
* 2014 update
* updated dependencies

20131230
* Worked on Python bindings

20131229
* updated dependencies

20131027
* Worked on Python bindings

20131017
* updated dependencies

20130909
* updated dependencies
* worked on libcthreads build support
* updated msvscpp files

20130530
* Worked on version resource support

20130528
* Worked on Python bindings
* Worked on version resource support

20130527
* Worked on Python bindings

20130526
* Worked on Python bindings

20130515
* updated dependencies
* moved examples to project site wiki
* added macosx files
* added Python bindings

20130421
* updated dependencies

20130413
* updated dependencies

20130405
* 2013 update
* updated dependencies
* Moved project to separate project site

20120402
* updated dependencies
* changes for libfdata update

20120319
* updated dependencies
* changes for libfdata update

20130310
* 2013 update
* fixed spec file
* updated dependencies
* updated msvscpp files

20121214
* worked on WEVT_TEMPLATE support

20121119
* worked on WEVT_TEMPLATE support

20121106
* code clean up

20121031
* bug fixes and code clean up

20121030
* bug fixes and code clean up

20121029
* worked on WEVT_TEMPLATE support

20121028
* worked on WEVT_TEMPLATE support
* update wrcinfo.1 man page

20121027
* worked on WEVT_TEMPLATE support
* code clean up
* wrcinfo debug output now skips already parsed resources

20121025
* updated dependencies
* worked on WEVT_TEMPLATE support

20121022
* small changes
* worked on WEVT_TEMPLATE support

20121021
* code clean up

20121017
* code clean up

20121015
* worked on MUI resource support
  - changed fallback language identifier to fallback language
  - added support for language
  - added support for non-set values

20121014
* worked on MUI resource support
* worked on version resource support
* worked on WEVT_TEMPLATE resource support

20121014
* added functionality for named resources
* worked on MUI resource support

20120925
* fixed bug introduced in libcdata update

20120924
* updated dependencies

20120912
* updated dependencies
* changes for libcdata update

20120627
* updated dependencies
* reversed library order in tools for mingw
* fix for deployment

20120506
* worked on event template resource support

20120505
* worked on named resource support
* worked on MUI resource support
* worked on event template resource support

20120501
* updated dependencies
* added debug dump of unknown resource data

20120422
* libfdatetime update
* libfvalue update

20120409
* updated dependencies

20120408
* updated dependencies

20120401
* moved version resource to language specific parsing
* worked on manifest support

20120322
* updated libuna

20120321
* small changes
* added check to see if language entry already exists and append strings

20120319
* worked on string resources

20120318
* worked on string resources

20120313
* worked on string resources

20120304
* added libfwnt support
* worked on multi language messages table
* fixed memory leak in tools

20120302
* fixed memory leak in library
* worked on multi language messages table

20120226
* worked on tooling

20120225
* 2012 update

20111124
* added libfcache
* updated libfdata, libexe

20111116
* added free of resource values

20111113
* worked on adding resources to API
* updated libexe

20111110
* worked on adding resources to API
* added libfvalue
* updated libexe

20111109
* worked on adding resources to API

20111108
* worked on adding resources to API

20111107
* worked on adding resources to API

20111106
* renamed libwmui to libwrc

20111101
* updated libcstring, libsystem, libuna, libbfio, libfdatetime, libexe
* added windows codepage 932 and 936 support

20111028
* worked on reading strings

20111027
* rename file to stream
* worked on reading MUI stream
* updated libexe

20111025
* updated m4 files
* updated libexe

20111024
* Initial version based on libexe



wrctools/wrcinfo.c
/*
 * Shows information obtained from a Windows Resource (RC) stream of the .rsrc section of a MZ, PE/COFF executable.
 *
 * Copyright (C) 2011-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "wrctools_getopt.h"
#include "wrctools_libcerror.h"
#include "wrctools_libclocale.h"
#include "wrctools_libcnotify.h"
#include "wrctools_libwrc.h"
#include "wrctools_output.h"
#include "wrctools_signal.h"
#include "wrctools_unused.h"

enum WRCINFO_MODES
{
	WRCINFO_MODE_RESOURCE_HIERARCHY,
	WRCINFO_MODE_OVERVIEW
};

info_handle_t *wrcinfo_info_handle = NULL;
int wrcinfo_abort                  = 0;

/* Signal handler for wrcinfo
 */
void wrcinfo_signal_handler(
      wrctools_signal_t signal WRCTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "wrcinfo_signal_handler";

	WRCTOOLS_UNREFERENCED_PARAMETER( signal )

	wrcinfo_abort = 1;

	if( wrcinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     wrcinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use wrcinfo to determine information about a Windows Resource (RC) stream of the .rsrc section of a MZ, PE/COFF executable.";

	wrctools_option_t options[ ] = {
		{ 'c', "codepage", "codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258" },
		{ 'h', NULL, "shows this help" },
		{ 'H', NULL, "shows the resource hierarchy" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libwrc_error_t *error               = NULL;
	system_character_t *option_codepage = NULL;
	system_character_t *source          = NULL;
	char *program                       = "wrcinfo";
	system_integer_t option             = 0;
	int number_of_options               = (int) ( sizeof( options ) / sizeof( wrctools_option_t ) );
	int option_mode                     = WRCINFO_MODE_OVERVIEW;
	int result                          = 0;
	int verbose                         = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "wrctools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( wrctools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	wrctools_output_version_fprint(
	 stdout,
	 program );

	if( wrctools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = wrctools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				wrctools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				wrctools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'H':
				option_mode = WRCINFO_MODE_RESOURCE_HIERARCHY;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				wrctools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		wrctools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libwrc_notify_set_stream(
	 stderr,
	 NULL );
	libwrc_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &wrcinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( option_codepage != NULL )
	{
		result = info_handle_set_ascii_codepage(
		          wrcinfo_info_handle,
		          option_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in info handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
	result = info_handle_open_input(
	          wrcinfo_info_handle,
	          source,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "No resource stream.\n" );

		goto on_error;
	}
	switch( option_mode )
	{
		case WRCINFO_MODE_RESOURCE_HIERARCHY:
			if( info_handle_resource_hierarchy_fprint(
			     wrcinfo_info_handle,
			     &error ) != 1 )
			{
				fprintf(
				 stderr,
				 "Unable to print resource hierarchy information.\n" );

				goto on_error;
			}
			break;

		case WRCINFO_MODE_OVERVIEW:
		default:
			if( info_handle_stream_fprint(
			     wrcinfo_info_handle,
			     &error ) != 1 )
			{
				fprintf(
				 stderr,
				 "Unable to print stream information.\n" );

				goto on_error;
			}
			break;
	}
	if( info_handle_close_input(
	     wrcinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &wrcinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( wrcinfo_info_handle != NULL )
	{
		info_handle_free(
		 &wrcinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

